feat: CartRegion link, definition + workflow

olivermrbl opened the pull request.

olivermrbl commented on packages/link-modules/src/loaders/container.ts:
```diff
@@ -1,5 +1,5 @@
-import { BaseRepository, getLinkRepository } from "@repositories"
-import { LinkService, getModuleService } from "@services"
+import { BaseRepository, getLinkRepository } from "../repositories"
```
question: @carlos-r-l-rodrigues, for some reason, the link module was throwing a `ModuleNotFoundError: Cannot find module '@repositories'` error upon booting the server in my integration tests. Same goes for `@services`.

Do you have any idea why?

carlos-r-l-rodrigues commented on packages/link-modules/src/definitions/cart-region.ts:
```diff
@@ -0,0 +1,58 @@
+    },
+    {
+      name: "cart_regions",
+    },
```
```suggestion
    {
      name: ["cart_region", "cart_regions"]
    },
```

carlos-r-l-rodrigues commented on packages/medusa/src/api-v2/store/carts/route.ts:
```diff
@@ -1,23 +1,32 @@
+    },
+  }
+
+  const [cart] = await remoteQuery(query)
```
### 
```suggestion
  const variables = { id: result[0].id }

  const query = {
    cart: {
      ...defaultStoreCartRemoteQueryObject,
    },
  }

  const [cart] = await remoteQuery(query, { cart: variables })
```

srindom reviewed (commented):
Should be changed to readonly link

sradevski commented on integration-tests/plugins/__tests__/cart/store/create-cart.spec.ts:
```diff
@@ -41,6 +46,38 @@ describe("POST /store/carts", () => {
+    )
+  })
+
+  it("should use any region", async () => {
```
Is the region selection deterministic when creating the cart? If so, it would be good to create 2 regions, and assert that eg. the first one is used. If not, we can still assert that the region is present in the response.

olivermrbl replied:
@sradevski, can I get you to expand on this? Since we don't have a default sorting, I have kept the test as is.

sradevski replied:
@olivermrbl yeah that is fine until sorting is deterministic, but I guess it's still good to check that the cart.region.id is set in the assertions.

olivermrbl replied:
Good push, this was an oversight from my side. Will add in

sradevski commented on packages/core-flows/src/definition/cart/steps/find-one-or-any-region.ts:
```diff
@@ -0,0 +1,27 @@
+    )
+
+    if (!data.regionId) {
+      const regions = await service.list({})
```
Making this deterministic by eg. sorting (maybe there already is some default sorting?) would be a bit better IMO.

adrien2p replied:
there is no default sorting indeed a default would be nice to have one indeed.

olivermrbl replied:
Adding a ticket for this @adrien2p

sradevski commented on packages/core-flows/src/definition/cart/steps/link-cart-region.ts:
```diff
@@ -0,0 +1,36 @@
+export const linkCartRegionStepId = "link-cart-region"
+export const linkCartRegionStep = createStep(
+  linkCartRegionStepId,
+  async (data: { regionId: string; cartId: string }, { container }) => {
```
Let's extract the type and use it below as well, so we don't have to do `data!`

adrien2p replied:
The type of data in the rollback function is inferred by the second argument of the step response, so here data will be `{ cartId, regionId } | undefined`. Now, we decided to make it also undefined, because it can happen that the rollback end up being called with no data in case the invoke function fail. I think here it is mostly an oversight, the first thing to do in that rollback is checking if that exists or return early, then data will not required `!`

sradevski commented on packages/medusa/src/api-v2/store/carts/route.ts:
```diff
@@ -1,23 +1,31 @@
-  res.status(200).json({ cart: result[0] })
+  const remoteQuery = req.scope.resolve("remoteQuery")
+
+  const variables = { id: result[0].id }
```
Is it worth asserting the result length in such cases?

adrien2p replied:
theoratically at this point, there will always be an item in the collection

sradevski reviewed (commented):
I am still not 100% sure how all of this ties together, so the comments are mostly on style/correctness of the code itself

olivermrbl commented on packages/link-modules/src/definitions/cart-region.ts:
```diff
@@ -0,0 +1,28 @@
+import { Modules } from "@medusajs/modules-sdk"
+import { ModuleJoinerConfig } from "@medusajs/types"
+
+export const CartRegion: ModuleJoinerConfig = {
```
comment: Since `region_id` lives on the Cart model, we don't need a link table to associate the two entities, so a read-only link is sufficient.

srindom commented on packages/link-modules/src/links.ts:
```diff
@@ -14,6 +14,12 @@ export const LINKS = {
     Modules.PRICING,
     "price_set_id"
   ),
+  CartRegion: composeLinkName(
```
q: should this be deleted now that it's read-only?

srindom replied:
I don't completely understand what this file does, but just saw the reference to a table name in the `composeLinkName` function. Since there's no join-table anymore, I thought it might not be needed anymore?

carlos-r-l-rodrigues replied:
this file is only a definition of all known links among Medusa modules.
We could bring these link names to the link definition itself, but there is the case where someone needs to replace a link module by a custom module and can use this to reference the original module name.

olivermrbl replied:
But @srindom is correct in that we don't need it for read-only links right?

carlos-r-l-rodrigues replied:
yes he is. I was just explaing what the file is haha. sorry.

carlos-r-l-rodrigues replied:
read only links do not generate a module. The definition is only used to compsoe the graph.

olivermrbl replied:
> yes he is. I was just explaing what the file is haha. sorry.

All good. I needed a refresher too, so thx! :D

olivermrbl commented:
Made a couple of additional changes in 1597853:
- Removed old create-cart workflow from the create cart endpoint, because it was messing with some test suites. This is theoretically a breaking change, as people could be using the it, but since the feature is in beta, and we've communicated that things are subject to change, I decided to get rid of it
- Added a transformer to add region currency to cart in case currency code is not passed to the workflow
- Moved workflow input type to `@medusajs/types`

carlos-r-l-rodrigues commented on packages/link-modules/src/definitions/cart-region.ts:
```diff
@@ -0,0 +1,28 @@
+      },
+    },
+    {
+      serviceName: Modules.REGION,
+      relationship: {
+        serviceName: Modules.CART,
+        primaryKey: "id",
+        foreignKey: "cart_id",
+        alias: "carts",
+        isList: true,
+      },
```
```suggestion
      serviceName: Modules.REGION,
      relationship: {
        serviceName: Modules.CART,
        primaryKey: "region_id",
        foreignKey: "id",
        alias: "carts",
        isList: true,
      },
```

carlos-r-l-rodrigues replied:
Maybe we can cover these configs on the tests, querying these new properties.

olivermrbl replied:
Something like this: https://github.com/medusajs/medusa/pull/6392/commits/2225f157fe49fdd428128dc778779d38164d15d9

carlos-r-l-rodrigues reviewed (approved):
LGTM
